lib/src/story/progress.rs

```diff
@@ -1730,658 +1379,89 @@ impl Story {
-            if let Ok(container) = first_content.clone().into_any().downcast::<Container>() {
-                flow_container = container;
-            } else {
-                break;
+    pub(crate) fn visit_container(&mut self, container: &Rc<Container>, at_start: bool) {
+        if !container.counting_at_start_only || at_start {
+            if container.visits_should_be_counted {
+                self.get_state_mut()
+                    .increment_visit_count_for_container(container);
             }
-        }
 
-        // Any initial tag objects count as the "main tags" associated with that
-        // story/knot/stitch
-        let mut in_tag = false;
-        let mut tags = Vec::new();
-
-        for content in &flow_container.content {
-            match content.as_ref().as_any().downcast_ref::<ControlCommand>() {
-                Some(command) => match command.command_type {
-                    CommandType::BeginTag => in_tag = true,
-                    CommandType::EndTag => in_tag = false,
-                    _ => {}
-                },
-                _ => {
-                    if in_tag {
-                        if let Some(string_value) = Value::get_string_value(content.as_ref()) {
-                            tags.push(string_value.string.clone());
-                        } else {
-                            return Err(
-                                StoryError::InvalidStoryState("Tag contained non-text content. Only plain text is allowed when using globalTags or TagsAtContentPath. If you want to evaluate dynamic content, you need to use story.Continue()".to_owned()),
-                            );
-                        }
-                    } else {
-                        break;
-                    }
-                }
+            if container.turn_index_should_be_counted {
+                self.get_state_mut()
+                    .record_turn_index_visit_to_container(container);
             }
         }
-
-        Ok(tags)
     }
 
-    fn content_at_path(&self, path: &Path) -> SearchResult {
-        self.main_content_container.content_at_path(path, 0, -1)
-    }
-
-    /// Gets a list of tags defined with '#' in the ink source that were seen
-    /// during the most recent [`cont`](Story::cont) call.
-    pub fn get_current_tags(&mut self) -> Result<Vec<String>, StoryError> {
-        self.if_async_we_cant("call currentTags since it's a work in progress")?;
-        Ok(self.get_state_mut().get_current_tags())
-    }
+    /// The vector of [`Choice`](crate::choice::Choice) objects available at
+    /// the current point in the `Story`. This vector will be
+    /// populated as the `Story` is stepped through with the
+    /// [`cont`](Story::cont) method.
+    /// Once [`can_continue`](Story::can_continue) becomes `false`, this
+    /// vector will be populated, and is usually (but not always) on the
+    /// final [`cont`](Story::cont) step.
+    pub fn get_current_choices(&self) -> Vec<Rc<Choice>> {
```

Comment:
Why not in `choices.rs`?

Reply:
For a downstream library user, I originally thought it would make sense to put the essential functions for progressing a story all together, so there's one main place in the docs to look for that. But it would also make sense for them to look in the `choices` section too, to find that function. Up to you which one you think is more intuitive.